File: src/main/js/formulas/math/SQRT.ts
import { NumException } from "../../errors/NumException";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class SQRT extends AbstractFormula {
  static SELF: SQRT = new SQRT();
  NAME = FormulaName.SQRT;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 1, this.NAME);
    const value = Converters.toNumber(Converters.first(this.collateralLookup(origin, values[0])));
    if (value < 0) {
      return new NumException(
        "Function SQRT parameter 1 value is negative. It should be greater than or equal to zero."
      );
    }
    return Math.sqrt(value);
  }
}

File: src/main/java/io/protobase/f7/formulas/math/SQRT.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.NumException;
import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class SQRT extends AbstractFormula {
  public static FormulaName NAME = FormulaName.SQRT;
  public static SQRT SELF = new SQRT();

  public SQRT() {
    super();
  }

  public SQRT(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 1, NAME);
    double value = Converters.toDouble(Converters.first(collateralLookup.apply(origin, values[0])));
    if (value < 0) {
      return new NumException("Function SQRT parameter 1 value is negative. It should be greater than or equal to zero.");
    }
    return Math.sqrt(value);
  }
}
